docs/dev/writing-code.md

```diff
@@ -192,18 +201,195 @@ TODO: document when to use future vs callback
+    *   Lives in Python module: `awscrt.io`
+    *   Source File: `awscrt/io.py`
+
+Creation goes like this:
+*   User's Python code creates an `EventLoopGroup`:
+    ```py
+    elg = EventLoopGroup()
+    ```
+*   `EventLoopGroup` initializer looks something like:
+    ```py
+    class EventLoopGroup:
+        def __init__(self, ...):
+            self._binding = _awscrt.event_loop_group_new(...)
+    ```
+*   `_awscrt.event_loop_group_new(...)` is Python calling down into C.
+    The C function looks something like:
+    ```C
+    PyObject *aws_py_event_loop_group_new(PyObject *self, PyObject *args) {
+        // ...parse arguments...
+
+        // allocate memory for binding struct
+        struct event_loop_group_binding *binding = aws_mem_calloc(...);
+
+        // create underlying implementation
+        binding->native = aws_event_loop_group_new(...);
+
+        // create PyCapsule which owns the binding struct.
+        // pass in "destructor" function that runs when
+        // PyCapsule is cleaned up by the garbage collector.
+        PyObject *capsule = PyCapsule_New(binding, on_capsule_destroyed_fn);
+        return capsule;
+    }
+    ```
+*   Things stay alive because:
+    *   The user's `elg` variable keeps the `EventLoopGroup` object alive.
+    *   Member variable `EventLoopGroup._binding` keeps the `PyCapsule` alive.
+    *   The `PyCapsule` keeps the `struct event_loop_group_binding` alive.
+    *   The `event_loop_group_binding.native` pointer is a "strong reference" that keeps
+        `struct aws_event_loop_group` alive.
+
+
+Destruction goes like this (it's actually more complex, we'll cover that later):
+*   When the user's Python code has no references to `elg`, the `EventLoopGroup` instance...
+*   The garbage collector cleans up the `EventLoopGroup`,
+    and the `PyCapsule` referenced by `EventLoopGroup._binding`.
+*   The `PyCapsule`'s destructor function runs, which looks something like:
+    ```C
+    static on_capsule_destroyed_fn(PyObject *capsule) {
+        struct event_loop_group_binding *binding = PyCapsule_GetPointer(capsule);
+
+        // release reference to underlying implementation
+        aws_event_loop_group_release(binding->native)
+
+        // free binding struct's memory
+        aws_mem_release(binding);
+    }
+    ```
+*   IF nothing else has a strong reference to `struct aws_event_loop_group`:
+    *   then it begins its shutdown process, and its memory is cleaned up when shutdown completes.
+*   ELSE something else has a strong reference to `struct aws_event_loop_group`:
+    *   so it won't begin its shutdown until the last reference is released.
+
+Note: In the past, the aws-c libs didn't have any reference counting for its C structs.
+You will still find older code in our Python bindings that tries to keep the entire dependency trees
+of Python objects alive via `Py_INCREF(x)` (TODO: remove needless complexity).
```

Comment:
Minor nitpick: I'd place the TODO on a separate line on a bullet:

~~~
Note: In the past, the aws-c libs didn't have any reference counting for its C structs.
You will still find older code in our Python bindings that tries to keep the entire dependency trees
of Python objects alive via Py_INCREF(x). You can't always look at existing code to see "the right way"
of doing things.
  * (TODO: Remove this needless complexity in CRT)
~~~

Reply:
I tried and it looked worse. I want this `Note` to be once block of text